core/src/smartcontracts/mod.rs

```diff
@@ -48,16 +35,44 @@ pub trait ValidQuery: Query {
     ///
     /// # Errors
     /// Concrete to each implementer
-    fn execute(&self, wsv: &WorldStateView) -> eyre::Result<Self::Output, QueryExecutionFailure>;
+    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, QueryExecutionFailure>;
 
     /// Executes query and maps it into value
     ///
     /// # Errors
     /// Concrete to each implementer
-    fn execute_into_value(
-        &self,
-        wsv: &WorldStateView,
-    ) -> eyre::Result<Value, QueryExecutionFailure> {
+    fn execute_into_value(&self, wsv: &WorldStateView) -> Result<Value, QueryExecutionFailure> {
         self.execute(wsv).map(Into::into)
     }
 }
+
+/// Context of expression evaluation
+#[derive(Clone)]
+pub struct Context<'wsv> {
+    values: BTreeMap<Name, Value>,
+    wsv: &'wsv WorldStateView,
+}
+
+impl<'a> Context<'a> {
+    /// Create new [`Self`]
+    pub fn new(wsv: &'a WorldStateView) -> Self {
+        Self {
+            values: BTreeMap::new(),
+            wsv,
+        }
+    }
+}
+
+impl iroha_data_model::evaluate::Context for Context<'_> {
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure> {
+        query.execute(self.wsv)
+    }
```

Comment:
I don't think it can be any other way. If this is not ok for you would you like to:

1. rename `Context` to something else?
2 entirely remove `Expression::Query`?

Reply:
> Context can execute Query? This doesn't make sense to me

I would rephrase it as "context can be queried", which sounds reasonable

Reply:
The difference is that Context now includes state of Iroha at the time of executing a query, not only identifiers found in the where clause of the expression.

Reply:
We could evaluate all expressions on the client side had it not been for the `Expression:Query`. Atm, it is the only expressions that needs `wsv`